File: leetcode-173_binary-search-tree-iterator.js
/**
 * 给定一个二叉搜索树, 实现 next 方法由小到大迭代返回二叉树节点
 */
var stack = []
var BSTIterator = function(root) {
    while (root) {
        stack.push(root)
        root = root.left
    }
};

BSTIterator.prototype.hasNext = function() {
    return stack.length > 0
};

BSTIterator.prototype.next = function() {
    let top = stack.pop()
    let ret = top.val
    if (top.right) { //栈顶元素就是最小节点
        top = top.right
        while (top) {
            stack.push(top)
            top = top.left
        }
    }
    return ret
};

File: leetcode-173_binary-search-tree-iterator.java
import java.util.Stack;

/**
 * 给定一个二叉搜索树, 实现 next 方法由小到大迭代返回二叉树节点
 */
class BSTIterator {

    private Stack<TreeNode> stack = new Stack<>();

    public BSTIterator(TreeNode root) {
        // 实例化对象时先沿着左儿子进栈
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
    }

    /** @return the next smallest number */
    public int next() {
        // 此时栈顶元素为最小
        TreeNode root = stack.pop();
        int ret = root.val;
        // 如果有右儿子，同样沿着左儿子进栈
        if (root.right != null) {
            root = root.right;
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
        }
        return ret;
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return stack.size() > 0;
    }
}
